### DISCUSSION TITLE
Drawing a texture during post processing

### DISCUSSION THREAD

**Discussion by TerranWorks** [bevyengine/bevy#11055] [Q&A] [answered]
TL;DR I want to draw crosshairs in the middle of the screen starting from the post_processing example but, I can't figure out to get my image into the BindGroupLayout

I've been trying to draw an image of crosshairs in the middle of the screen on top of my 3d scene, but this has turned out to be really difficult. I initially wanted to use a sprite component, but there were issues with it being drawn behind the 3d scene and also rotating along with the player, so I decided to try a making it shader instead (I eventually want to change the color of the crosshairs to make it visible on different color backgrounds, so I would probably make it a shader eventually anyway). Because I want to sample the screen texture I believe I need to follow the post_processing example and I have been able to integrate that into my game with the chromatic aberration shader, but I can't figure out how to add my texture to the bind group! I've found the ``create_texture_with_data`` method from ``RenderDevice`` and I can see that I can the ``Image`` in my ``Handle<Image>`` has the ``pub data: Vec<u8>`` field which makes me think I'm supposed to get the image from the handle, take the data out, and use ``create_texture_with_data`` to create a Texture which I can then pass along during the ``create_bind_group`` step of the post_processing example like is being done with ``post_process.source`` but, I'm not confident I'm even on the right track. Am I close? Any guidance would be appreciated

**Comment by JMS55** [comment, marked as answer]
You'll want to use Res<RenderAssets<Image>> from the render world in order to pull out a GpuImage, which has a texture view you can add to your bind group. 

Don't have time to explain in further detail, but hopefully that puts you on the right track.

**Reply by TerranWorks** [reply]
Thanks for your reply. Your suggestion led me to look at the ``compute_shader_game_of_life`` example which gave me what I needed. The only problem I'm still having is  ``gpu_images.get `` sometimes returning None in the following snippet:
```rust
    fn prepare_bind_group(
        mut commands: Commands,
        crosshairs_pipeline: Res<CrosshairsPipeline>,
        gpu_images: Res<RenderAssets<Image>>,
        crosshairs_image: Res<CrosshairsImage>,
        render_device: Res<RenderDevice>,
    ) {
        // Convert Handle<Image> in main world into GpuImage in render world
        let gpu_image = gpu_images.get(&crosshairs_image.0).unwrap();
        // ...
    }
```
It spontaneously started working, but I think it was happening because my ``CrosshairsImage`` asset hadn't finished loading in when the ``prepare_bind_group`` system was called. I did some testing to make sure the path was correct by spawning a sprite and then when I changed to code back it started working so I'm not sure what changed. Is there a way to ensure the image has loaded before trying to get the GpuImage?

**Reply by TerranWorks** [reply]
It stopped working again pretty quickly. I tried using AssetEvents to detect when the image is Added and create the BindGroup at that time, while the run method of my ViewNode just returns early if the BindGroup isn't available yet. Unfortunately, I couldn't figure out how to get the AssetEvent from the normal world while also working in the RenderWorld.

**Reply by JMS55** [reply]
You're mostly on the right track. Yes, the asset might not be ready (on the CPU or GPU). What I would do (this is what bevy itself does) is in prepare_bind_groups(), if RenderAssets::get() returns None, do nothing. If it returns Some, then create your bind group and add it as a component to the view (camera) entity. Then, in your ViewNode, simply put your BindGroupWrapperThing component as part of the view query. The node will only run if the component exists, which will only happen if the image was ready. See bevy_pbr's source code for various examples of this (SSAO for instance).

**Reply by TerranWorks** [reply]
That works really well. I have to say, I've been playing around with bevy for a couple of weeks now and am continuously impressed with how many clever solutions are packed into it. I'm looking forward to getting on with my project. Thanks!
